### ISSUE TITLE
No EOS token?

### ISSUE THREAD

**Issue by SmartWashingMachine** [jzhang533/paddleocr-vl-sft#1]
Hi, thanks for the example. But is this part right?

```
            caption = item["text"]
            messages = [{"role": "user", "content": self.task_prompt}]
            text_prompt = self.processor.apply_chat_template(
                messages,
                tokenize=False,
                add_generation_prompt=True
            )
            full_text = text_prompt + caption
```

If we print out the full_text we get:
`<|begin_of_sentence|>User: <|IMAGE_START|><|IMAGE_PLACEHOLDER|><|IMAGE_END|>OCR:
Assistant: 素直にあやまるしかあやまるわよ！`

But no `<|end_of_sentence|>` token? So model would learn to repeat tokens indefinitely?

Looking at their [chat template](https://huggingface.co/PaddlePaddle/PaddleOCR-VL/blob/main/chat_template.jinja) it seems that `<|end_of_sentence|>`  was not added because there's no assistant role message.

So maybe we need this instead:

```
                messages = [{"role": "user", "content": self.task_prompt}, {"role": "assistant", "content": caption, }]
                text_prompt = self.processor.apply_chat_template(
                    messages,
                    tokenize=False,
                    add_generation_prompt=False
                )
                full_text = text_prompt
```

Which prints:
`<|begin_of_sentence|>User: <|IMAGE_START|><|IMAGE_PLACEHOLDER|><|IMAGE_END|>OCR:
Assistant: 素直にあやまるしかあやまるわよ！<|end_of_sentence|>`

Thank you again for this example. It is very helpful.

**Comment by jzhang533** [comment]
@SmartWashingMachine thanks a lot for sharing your findings!

I dig more into SFT PaddleOCR-VL-0.9B recently, the code I used is in wip branch in this repository, currently.

Some key points I found are

1. use completion_only loss, i.e.: "train on responses only“ as you did, I have a similar approach here: https://github.com/jzhang533/paddleocr-vl-sft/blob/wip/custom_collator.py
2. transformers/trl respect to chat_template.json and tokenizer_config.json, so I think it use `<|end_of_sentence|>` as EOS, but the pretrained model actually is using `</s>` token. For me, I didn't change any config of the released model, instead, when inference after SFT, I am using both `<|end_of_sentence|>` and `</s>` token as EOS, as shown here : https://github.com/jzhang533/paddleocr-vl-sft/blob/wip/use_transformers.py#L70 . It works for me.
3. don‘t use truncation when training, otherwise it may truncate vision tokens, this will cause training crash.

BTW: I shared a synthetic dataset on huggingface: https://huggingface.co/datasets/jzhang533/manga-synthetic, planned to consolidate the findings, and share a tutorial using this dataset on how to SFT this model using transformers and trl library.

**Comment by Sunting78** [comment]
Thank you for your attention. In PaddleOCR-VL, <|end_of_sentence|> is actually the sep token, and </s> is the eos. It seems that the setting of eos as <|end_of_sentence|> is due to the way data is organized during training in transformers. Your addition of eos_token_id=100272 is helpful in this training code. It is also correct to mask the image tokens you mentioned in the training code.
